IGNITE-21899 added GridInt utils

FewJuho opened the pull request.

FewJuho committed:
added GridInt utils

new ver 2.16

added ignite-indexing

FewJuho committed:
migrator GridIntList

FewJuho committed:
Merge branch 'apache:master' into ignite-21899

FewJuho committed:
added README

FewJuho committed:
Merge branch 'ignite-21899' of github.com:FewJuho/ignite-teamcity-bot into ignite-21899
.

timoninmaxim commented on build.gradle:
```diff
@@ -46,7 +46,7 @@ allprojects {
         jettyVer = '9.4.12.v20180830'
 
-        ignVer = '2.14.0'
+        ignVer = '2.16.0'
```
2.17.0 is actual version

timoninmaxim commented on migrator/src/main/java/migrate/GridIntListMigrator.java:
```diff
@@ -0,0 +1,532 @@
+            }
+            catch (Throwable t) {
+                if (verbose) {
+                    log.warn("New keys type {} is not on classpath, will fallback to int[] (cause: {})",
```
At which circumstance we should proceed if class not found?

timoninmaxim commented on migrator/src/main/java/migrate/GridIntListMigrator.java:
```diff
@@ -0,0 +1,532 @@
+     * --report <N>      progress log interval
+     * --workDir <path>  path to work/ directory (overrides IGNITE_WORK_DIR)
+     */
+    static final class Args {
```
Let's create separate files for each package-private class: Args, Transformer, TransformResult.

timoninmaxim commented on migrator/src/main/java/migrate/GridIntListMigrator.java:
```diff
@@ -0,0 +1,532 @@
+        try (Ignite ig = Ignition.start(cfg)) {
+            ig.cluster().state(ClusterState.ACTIVE);
+
+            Collection<String> cacheNames = new ArrayList<>(ig.cacheNames());
```
Do we need scan all caches?

FewJuho replied:
Seems good to scan all caches to prevent cache data loss. There's CLI command --cache to specify caches' names to scan.

timoninmaxim commented on migrator/src/main/java/migrate/GridIntListMigrator.java:
```diff
@@ -0,0 +1,532 @@
+ * ./gradlew -p migrator run --args="--apply --report 500"        # apply to all caches
+ */
+
+public final class GridIntListMigrator {
```
I see we have DbMigrations class with caches/data migrations. Can we built this code in the class?

timoninmaxim replied:
Most of migrations are one-shot tasks, see DbMigrations#applyMigration - it skips a migration if it had been already applied. Is there other reason to keep this migration in a separate module?

FewJuho committed:
review fixes

FewJuho committed:
package name fixed

FewJuho committed:
IGNITE-26338 Upgrade java version to 11 (#201)

FewJuho committed:
Merge branch 'master' into ignite-21899

timoninmaxim commented on migrator/src/main/java/migrate/Transformer.java:
```diff
@@ -0,0 +1,287 @@
+                int[] res = new int[g.size()];
+
+                for (int i = 0; i < res.length; i++)
+                    res[i] = g.get(i);
```
Can we use System.arrayCopy instead?

timoninmaxim commented on migrator/src/main/java/migrate/Transformer.java:
```diff
@@ -0,0 +1,287 @@
+        }
+        catch (Throwable t) {
+            if (verbose)
+                log.info("Deserialize fallback: {}", t.toString());
```
At which case we failed here?

timoninmaxim commented on migrator/src/main/java/migrate/Transformer.java:
```diff
@@ -0,0 +1,287 @@
+     * or falls back to int[] if the class is not on the classpath.
+     */
+    private Object buildNewKeys(int[] ints) {
+        if (newKeysCachedConstruct != null) {
```
it never null

FewJuho committed:
review fixes: package and arrayCopy

FewJuho committed:
review fixes: package and arrayCopy

FewJuho committed:
review fixes: DbMigrations

fixed migrator module build file

fixed build

updated DbMigrations

added GridIntList migration into DbMigration class

fixed reportEvery to Int

dedupl

FewJuho committed:
Merge branch 'ignite-21899' of github.com:FewJuho/ignite-teamcity-bot into ignite-21899

dspavlov commented:
resolved by #206 , thank you for your contribution
